### PR TITLE
ENH: Switch to Codecov

### PR DESCRIPTION
**Pull request by ant0nsc**
Retire the old manual code coverage reporting and switch to Codecov.io

### CODE REVIEW

File: .github/workflows/build-test-pr.yml [hunk 1]
```diff
@@ -204,11 +204,12 @@ jobs:
...
-      - name: Upload coverage artifacts
-        uses: ./.github/actions/upload_coverage_artifacts
+      - name: Upload coverage report to Codecov
         if: ${{ matrix.packageName == '*.whl' }}
-        with:
-          folder: ${{ matrix.folder }}
+        run: |
+          curl -Os https://uploader.codecov.io/latest/linux/codecov
+          chmod +x codecov
+          ./codecov
```
Comment (fepegar) [inline .github/workflows/build-test-pr.yml:212]: Can't it be pip-installed?

File: hi-ml-azure/testazure/testazure/test_azure_util.py [hunk 2]
```diff
@@ -1372,17 +1372,17 @@ def test_replace_directory(tmp_path: Path) -> None:
...
+    assert (output_dir / file_name).exists()
+    new_output_dir = output_dir.parent / "more_test_outputs"
 
-replace_directory(output_dir, new_output_dir)
+    replace_directory(output_dir, new_output_dir)
 
-assert not output_dir.exists()
-assert (new_output_dir / file_name).exists()
+    assert not output_dir.exists()
+    assert (new_output_dir / file_name).exists()
```
Comment (fepegar) [inline hi-ml-azure/testazure/testazure/test_azure_util.py:1385]: Are these changes related to this PR?

**Comment by fepegar** [comment]
Is the [Codecov message](https://github.com/microsoft/hi-ml/pull/340#issuecomment-1125900768) expected? Is it because it doesn't have any previous info?

File: .github/workflows/build-test-pr.yml (hunk 1 above)
Comment (ant0nsc) [reply .github/workflows/build-test-pr.yml:212]: that's the suggested use in the codecov documentation, just following that.

File: .github/workflows/build-test-pr.yml [hunk 3]
```diff
@@ -259,92 +260,20 @@ jobs:
           make pip_local
           make pytest_and_coverage
 
+      - name: Upload coverage report to Codecov
```
Comment (mebristo) [inline .github/workflows/build-test-pr.yml:263]: Why do we upload these to Codecov?

File: hi-ml-azure/testazure/testazure/test_azure_util.py (hunk 2 above)
Comment (mebristo) [reply hi-ml-azure/testazure/testazure/test_azure_util.py:1385]: This test has been bugging me too recently - weird!

File: .github/workflows/build-test-pr.yml (hunk 1 above)
Comment (fepegar) [reply .github/workflows/build-test-pr.yml:212]: Shouldn't we use the GHA instead?

https://docs.codecov.com/docs/codecov-uploader#introduction

File: hi-ml-histopathology/Makefile [hunk 4]
```diff
@@ -86,10 +86,9 @@ call_pytest_fast:
 # install test requirements and run pytest fast subset
 pytest_fast: pip_test call_pytest_fast
 
-# run pytest with coverage on package, and format coverage output as a text file, assuming test requirements already installed
+# run pytest with coverage on package
 call_pytest_and_coverage:
```
[the remark this answers is not in the document]
Comment (mebristo) [reply hi-ml-histopathology/Makefile:90]: [coverage(.py)](https://coverage.readthedocs.io/en/6.3.3/) is the library that the plugin pytest-cov uses to produce reports. So rather I should have asked if this is the same pytest-cov command

[the remark this answers is not in the document]
Comment (fepegar) [reply hi-ml-histopathology/Makefile:90]: Here's a minimal repo by codecov, showcasing how to use it in a Python repo: https://github.com/codecov/python-standard

File: .github/workflows/build-test-pr.yml (hunk 3 above)
Comment (ant0nsc) [reply .github/workflows/build-test-pr.yml:263]: because the tool is cloud driven. Just following the suggested process for codecov.

File: .github/workflows/build-test-pr.yml (hunk 1 above)
Comment (fepegar) [reply .github/workflows/build-test-pr.yml:212]: Can we just double-check whether the GHA is the recommended setup?

**Review by mebristo — APPROVED** [review]

File: .github/workflows/build-test-pr.yml (hunk 1 above)
Comment (ant0nsc) [reply .github/workflows/build-test-pr.yml]: There's no clear recommendation, but I've switched to using the Github action. A lot nicer now, in my opinion.

Comment (fepegar) [reply .github/workflows/build-test-pr.yml]: I agree!

**Comment by fepegar** [comment]
> Is the https://github.com/microsoft/hi-ml/pull/340#issuecomment-1125900768 expected? Is it because it doesn't have any previous info?

To reply to my own message: of course, it makes sense. There's no coverage for main yet!

**Review by fepegar — APPROVED** [review]

**Comment by fepegar** [comment]
Thanks, @ant0nsc! I'm excited about this change. Yes, my excitement bar is a bit low.
